komodo/check_up_to_date_pypi.py

```diff
@@ -0,0 +1,167 @@
+    ]
+
+
+def get_python_requirement(sources: list):
+    """
+    Goes through the different sources of the package and returns the first
+    python requirement
+    :param sources:
+    :return: Python requirement or empty string if there is no requirement
+    """
+    for source in sources:
+        if source.get("yanked", False):
+            continue
+        required_python = source.get("requires_python", "")
+        if required_python:
+            return required_python
+    return ""
+
+
+def get_compatible_versions(releases: dict, python_version):
+    compatible_versions = []
+    for version_str, build_info in releases.items():
+        package_version = get_version.parse(version_str)
+        if package_version.is_prerelease:
+            continue
+        required_python = SpecifierSet(get_python_requirement(build_info))
+        if python_version in required_python:
+            compatible_versions.append(package_version)
+    return compatible_versions
+
+
+def evaluate_response(response: dict, python_version: str):
+    python_requirement = response["info"]["requires_python"]
+    if not python_requirement or python_version in SpecifierSet(python_requirement):
+        return get_version.parse(response["info"]["version"])
+    return max(get_compatible_versions(response["releases"], python_version))
+
+
+def get_pypi_packages(release: dict, repository: dict) -> list:
+    pypi_packages = []
+    for package, version in release.items():
+        if repository[package][version].get("source", None) == "pypi":
+            pypi_packages.append(package)
+    return pypi_packages
+
+
+def get_upgrade_proposals(release_file, repository_file, python_version, propose_upgrade):
+    yaml = yaml_parser()
+    releases = get_release(yaml, release_file)
+    repository = get_release(yaml, repository_file)
+
+    pypi_packages = get_pypi_packages(releases, repository)
+    pypi_response = get_packages_pypi(pypi_packages)
+
+    upgrade_proposals = {}
+    for name, response in pypi_response:
+        komodo_version = strip_version(releases[name])
+        if LATEST_PACKAGE_ALIAS in komodo_version:
+            continue
+        komodo_version = get_version.parse(komodo_version)
+        if response.ok:
+            pypi_latest_version = evaluate_response(response.json(), python_version)
+        else:
+            raise ValueError(
+                f"Response returned non valid return code: {response.reason}"
+            )
+        if pypi_latest_version != komodo_version:
+            upgrade_proposals[name] = {"previous": str(komodo_version), "suggested": str(pypi_latest_version)}
+
+    if propose_upgrade and upgrade_proposals:
+        for name, version in upgrade_proposals.items():
+            suggested_version = version["suggested"]
+            if suggested_version not in repository[name]:
+                repository[name].update(
+                    {suggested_version: copy.deepcopy(repository[name][releases[name]])}
```

Comment:
If I have understood this correct, then maybe this reads a little bit better? :thinking: 
```suggestion
                    {suggested_version: copy.deepcopy(repository[name][version["previous"]])}
```

Reply:
I agree that would be better, but the version there has been sanitized (to remove `+py3` or what we might find appropriate to add), so will not fit directly. Will try to improve readability.

Reply:
And at the same time, add a test for that `+py3` thing